File: litedb-java/src/main/java/com/litedb/cluster/ClusterSmoke.java
package com.litedb.cluster;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/** Headless end-to-end test: 3 real NodeServer processes driven by ClusterClient. RF-aware. */
public class ClusterSmoke {

    static Map<String, String> leadersMap(ClusterClient client) {
        Map<String, String> m = new TreeMap<>();
        for (Map<String, Object> node : client.status()) {
            if (!Boolean.TRUE.equals(node.get("alive"))) continue;
            @SuppressWarnings("unchecked")
            List<Object> shards = (List<Object>) node.get("shards");
            for (Object so : shards) {
                @SuppressWarnings("unchecked")
                Map<String, Object> sh = (Map<String, Object>) so;
                if ("leader".equals(sh.get("role"))) m.put((String) sh.get("group"), (String) sh.get("node"));
            }
        }
        return m;
    }

    static void waitUntil(java.util.function.BooleanSupplier cond, long timeoutMs, String what) throws Exception {
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (System.currentTimeMillis() < deadline) {
            if (cond.getAsBoolean()) return;
            Thread.sleep(200);
        }
        throw new AssertionError("timeout waiting for " + what);
    }

    static void rmrf(String dir) throws Exception {
        Path p = Paths.get(dir);
        if (!Files.exists(p)) return;
        Files.walk(p).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
    }

    public static void main(String[] args) throws Exception {
        int rf = ClusterConfig.replicationFactor();
        Partitioner part = ClusterConfig.makePartitioner();
        rmrf(ClusterConfig.dataRoot());

        String cp = System.getProperty("java.class.path");
        String javaBin = System.getProperty("java.home") + "/bin/java";
        Map<String, Process> procs = new LinkedHashMap<>();
        for (String nid : ClusterConfig.nodeIds()) {
            ProcessBuilder pb = new ProcessBuilder(javaBin, "-cp", cp, "com.litedb.cluster.NodeServer", nid);
            pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
            pb.redirectError(ProcessBuilder.Redirect.DISCARD);
            procs.put(nid, pb.start());
        }
        try {
            ClusterClient client = new ClusterClient();
            System.out.println("Replication factor = " + rf + " (each shard on " + rf + " of "
                    + part.nodes().size() + " nodes)");
            waitUntil(() -> leadersMap(client).size() == 6, 20000, "6 leaders");
            Map<String, String> lm = leadersMap(client);
            Map<String, Integer> spread = new TreeMap<>();
            for (String n : lm.values()) spread.merge(n, 1, Integer::sum);
            System.out.println("  leaders per node: " + spread + "  (multi-raft spreads leadership)");

            Map<String, Integer> hosted = new TreeMap<>();
            int maxHosted = 0;
            for (String n : part.nodes()) { int c = part.shardsOn(n).size(); hosted.put(n, c); maxHosted = Math.max(maxHosted, c); }
            System.out.println("  shards hosted per node: " + hosted);
            if (rf < part.nodes().size() && maxHosted >= part.shardIds().size())
                throw new AssertionError("RF<N should leave some shard off some node");

            System.out.println("Writing 12 keys (routed by consistent hashing)...");
            for (int i = 0; i < 12; i++) {
                if (!Boolean.TRUE.equals(client.put("key" + i, "val" + i).get("ok")))
                    throw new AssertionError("put key" + i);
            }
            boolean allRead = true;
            for (int i = 0; i < 12; i++) allRead &= ("val" + i).equals(client.get("key" + i));
            System.out.println("  all 12 reads correct -> " + allRead);
            if (!allRead) throw new AssertionError("reads");

            // cross-shard txn
            String a = "key0", b = null;
            for (int i = 1; i < 12; i++) if (!part.shardFor("key" + i).equals(part.shardFor("key0"))) { b = "key" + i; break; }
            Map<String, Object> w = new LinkedHashMap<>();
            w.put(a, "alice=900"); w.put(b, "bob=100");
            Map<String, Object> tr = client.txn(w, null);
            System.out.println("  cross-shard 2PC -> ok=" + tr.get("ok") + " shards=" + tr.get("shards"));
            if (!Boolean.TRUE.equals(tr.get("ok"))) throw new AssertionError("2pc");
            if (!"alice=900".equals(client.get(a)) || !"bob=100".equals(client.get(b))) throw new AssertionError("2pc data");

            // snapshot isolation
            Long snap = client.begin();
            client.put(a, "alice=CHANGED");
            String old = client.get(a, snap), now = client.get(a);
            System.out.println("  snapshot=" + old + "  latest=" + now);
            if (!"alice=900".equals(old) || !"alice=CHANGED".equals(now)) throw new AssertionError("snapshot isolation");

            if (rf >= 3) {
                String victim = null; int best = -1;
                for (Map.Entry<String, Integer> e : spread.entrySet()) if (e.getValue() > best) { best = e.getValue(); victim = e.getKey(); }
                System.out.println("Killing " + victim + " (led " + best + " shards)...");
                procs.get(victim).destroyForcibly().waitFor();
                final String dead = victim;
                waitUntil(() -> { Map<String, String> m = leadersMap(client); return m.size() == 6 && !m.containsValue(dead); }, 20000, "failover");
                System.out.println("  re-elected onto survivors");
                if (!Boolean.TRUE.equals(client.put("after-failover", "ok").get("ok"))) throw new AssertionError("post-failover write");
                if (!"alice=CHANGED".equals(client.get(a))) throw new AssertionError("data lost");
                System.out.println("  write + read after failover OK; pre-crash data intact");
            } else {
                System.out.println("(RF=" + rf + " < 3 → single node loss removes a shard's majority; failover not exercised.)");
            }
            System.out.println("\nCLUSTER OK at RF=" + rf + " (Java): partitioning, multi-raft, RF-agnostic routing, "
                    + "cross-shard 2PC, snapshot isolation" + (rf >= 3 ? ", live failover." : "."));
        } finally {
            for (Process p : procs.values()) p.destroyForcibly();
            for (Process p : procs.values()) p.waitFor();
        }
        System.exit(0);
    }
}

File: litedb-python/cluster_smoke.py
"""Scratch test: 3 real node processes, driven by ClusterClient. Phases 4-7 end to end."""
import shutil
import subprocess
import sys
import time

from cluster_client import ClusterClient
from cluster_config import DATA_ROOT, NODES, make_partitioner


def wait_until(fn, timeout=10.0, what=""):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if fn():
            return True
        time.sleep(0.2)
    raise AssertionError(f"timeout waiting for {what}")


def leaders_map(client):
    """shard -> leader node, as reported by status."""
    m = {}
    for node in client.status():
        if not node.get("alive"):
            continue
        for sh in node.get("shards", []):
            if sh["role"] == "leader":
                m[sh["group"]] = sh["node"]
    return m


def main():
    shutil.rmtree(DATA_ROOT, ignore_errors=True)
    procs = {}
    for nid in NODES:
        procs[nid] = subprocess.Popen([sys.executable, "node.py", nid],
                                      stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    try:
        client = ClusterClient()
        part = make_partitioner()

        print("Waiting for all 6 shards to elect leaders...")
        wait_until(lambda: len(leaders_map(client)) == 6, timeout=15, what="6 leaders")
        lm = leaders_map(client)
        from collections import Counter
        spread = Counter(lm.values())
        print(f"  leaders per node: {dict(spread)}  (multi-raft spreads leadership)")
        assert len(spread) >= 2, "leadership should spread across nodes"

        print("\nWriting 12 keys (routed to their shards by consistent hashing):")
        for i in range(12):
            r = client.put(f"key{i}", f"val{i}")
            assert r.get("ok"), f"put key{i} failed: {r}"
        by_shard = {}
        for i in range(12):
            by_shard.setdefault(part.shard_for(f"key{i}"), []).append(f"key{i}")
        print(f"  keys landed across {len(by_shard)} shards: "
              f"{ {s: len(ks) for s, ks in sorted(by_shard.items())} }")

        print("\nReading them all back (contacting random nodes; they forward to leaders):")
        ok = all(client.get(f"key{i}") == f"val{i}" for i in range(12))
        print(f"  all 12 reads correct -> {ok}")
        assert ok

        print("\nCross-shard transaction (atomic transfer across two shards):")
        # pick two keys on different shards
        a, b = "key0", None
        for i in range(1, 12):
            if part.shard_for(f"key{i}") != part.shard_for("key0"):
                b = f"key{i}"
                break
        print(f"  {a} (shard {part.shard_for(a)})  +  {b} (shard {part.shard_for(b)})")
        r = client.txn({a: "alice=900", b: "bob=100"})
        print(f"  2PC result -> ok={r.get('ok')} shards={sorted(r.get('shards', []))}")
        assert r.get("ok") and len(r.get("shards", [])) == 2
        assert client.get(a) == "alice=900" and client.get(b) == "bob=100"

        print("\nSnapshot isolation across the cluster:")
        snap = client.begin()
        client.put(a, "alice=CHANGED")
        old = client.get(a, read_ts=snap)
        new = client.get(a)
        print(f"  @snapshot={old!r}  latest={new!r}")
        assert old == "alice=900" and new == "alice=CHANGED"

        print("\nKilling a node and verifying failover + continued service:")
        victim = max(spread, key=spread.get)  # node leading the most shards
        print(f"  killing {victim} (was leading {spread[victim]} shards)")
        procs[victim].kill()
        procs[victim].wait()
        wait_until(lambda: len(leaders_map(client)) == 6, timeout=15,
                   what="re-election of victim's shards onto survivors")
        lm2 = leaders_map(client)
        assert victim not in set(lm2.values()), "dead node still shown as leader"
        print(f"  re-elected: leaders now on {sorted(set(lm2.values()))}")

        print("\nWrites still succeed after failover:")
        r = client.put("after-failover", "ok")
        assert r.get("ok"), r
        assert client.get("after-failover") == "ok"
        assert client.get(a) == "alice=CHANGED", "data from before the crash survived"
        print("  write + read after failover OK; pre-crash data intact")

        print("\nPHASES 4-7 OK: partitioning, multi-raft, routing, cross-shard 2PC, "
              "snapshot isolation, and live failover — across 3 real processes.")
    finally:
        for p in procs.values():
            if p.poll() is None:
                p.kill()
        for p in procs.values():
            p.wait()


if __name__ == "__main__":
    main()
